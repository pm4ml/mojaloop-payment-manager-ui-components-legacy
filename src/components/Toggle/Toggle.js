import './Toggle.scss';

import PropTypes from 'prop-types';
import React, { PureComponent } from 'react';

import * as utils from '../../utils/common';

class Toggle extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      checked: this.props.checked,
    };
    this.onClick = this.onClick.bind(this);
    this.onChange = this.onChange.bind(this);
    this.onFocus = this.onFocus.bind(this);
    this.onBlur = this.onBlur.bind(this);
    this.testKey = this.testKey.bind(this);
  }
  componentDidUpdate(prevProps) {
    if (this.props.checked !== prevProps.checked) {
      // eslint-disable-next-line
      this.setState({ checked: this.props.checked });
    }
  }

  onChange(e) {
    e.preventDefault();
    e.stopPropagation();
    if (this.props.disabled) {
      return;
    }

    const checked = !this.state.checked;
    this.setState({ checked });
    if (this.props.onChange) {
      this.props.onChange(checked);
    }
  }
  onBlur(e) {
    if (this.props.onBlur) {
      this.props.onBlur(e);
    }
  }
  onClick(e) {
    e.preventDefault();
    e.stopPropagation();
    if (this.props.onClick) {
      this.props.onClick(e);
    }
  }
  onFocus(e) {
    if (this.props.onFocus) {
      this.props.onFocus(e);
    }
  }
  testKey(e) {
    if (e.keyCode === 9) {
      e.stopPropagation();
      e.preventDefault();
      utils.focusNextFocusableElement(this.input, !e.shiftKey);
      return;
    }
    if (e.keyCode === 13) {
      e.stopPropagation();
      e.preventDefault();
      this.onChange(e);
    }
  }
  render() {
    const { checked } = this.state;
    const { style, className, id, label, disabled } = this.props;
    const toggleClassName = utils.composeClassNames(['mb-input', 'input-toggle', className]);
    const labelClassName = utils.composeClassNames([!label && 'input-toggle__label--no-margin']);
    return (
      <div className={toggleClassName} style={style}>
        <input
          ref={input => {
            this.input = input;
          }}
          type="checkbox"
          id={id}
          className="input-toggle__input"
          onClick={this.onClick}
          onBlur={this.onBlur}
          onFocus={this.onFocus}
          onKeyDown={this.testKey}
          onChange={e => {
            e.preventDefault();
            e.stopPropagation();
          }}
          checked={checked}
          disabled={disabled}
        />
        {/* eslint-disable-next-line */}
        <label htmlFor={id} className={labelClassName} onClick={this.onChange} role="button">
          {label}
        </label>
      </div>
    );
  }
}
Toggle.propTypes = {
  style: PropTypes.shape(),
  className: PropTypes.string,
  id: PropTypes.string,
  label: PropTypes.string,
  onBlur: PropTypes.func,
  onClick: PropTypes.func,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  checked: PropTypes.bool,
  disabled: PropTypes.bool,
};
Toggle.defaultProps = {
  style: undefined,
  className: undefined,
  id: undefined,
  label: undefined,
  onBlur: undefined,
  onClick: undefined,
  onChange: undefined,
  onFocus: undefined,
  checked: false,
  disabled: false,
};
export default Toggle;
