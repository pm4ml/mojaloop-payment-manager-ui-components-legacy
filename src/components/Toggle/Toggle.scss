@import '../../assets/styles/vars/colors.scss';
@import '../../assets/styles/vars/shadow.scss';

.input-toggle {
  display: flex;
  align-items: center;
}
.input-toggle__input {
  width: 0px;
  height: 0px;
  appearance: none;
  opacity: 0;
  position: relative;
  margin: 0px;
  padding: 0px;

  &:focus {
    + label {
      &:before {
        box-shadow: $shadow-size rgba(0, 175, 255, 0.5);
      }
    }
  }

  + label {
    max-width: 100%;
    display: flex;
    position: relative;
    cursor: pointer;
    align-items: center;
    font-size: 14px;
    user-select: none;
    height: 22px;

    &:hover {
      &:before {
        box-shadow: inset 0px 1px 2px 0px rgba(0, 0, 0, 0.1), $shadow-size rgba(0, 0, 0, 0.05);
      }
    }
    &:before {
      border-radius: 12px;
      box-sizing: initial;
      flex: 0 0 40px;
      display: inline-block;
      margin-right: 5px;
      background: #ddd;
      content: '';
      width: 20px;
      height: 20px;
      box-shadow: inset 0px 1px 2px 0px rgba(0, 0, 0, 0.1), 0px 0px 0px 0px rgba(0, 0, 0, 0.13);
      transition: all 0.2s ease-out;
      flex-basis: 40px;
      border: none;
      height: 20px;
      width: 40px;
    }
    &:after {
      content: '';
      box-sizing: content-box;
      box-sizing: initial;
      position: absolute;
      left: 2px;
      border-radius: 10px;
      top: 3px;
      background: #fff;
      width: 16px;
      height: 16px;
      background: #fff;
      transition: all 0.2s ease-out;
      box-shadow: 0px 1px 1px 0px rgba(0, 0, 0, 0.2);
    }

    &.input-toggle__label--no-margin {
      &:before {
        margin-right: 0px;
      }
    }
  }

  &:disabled {
    + label {
      color: #999;
      cursor: default;
      &:before {
        background: #eee;
        box-shadow: none;
      }
    }
    &:checked {
      + label {
        &:before {
          background: #60d2fd;
        }
      }
    }
  }

  &:checked {
    + label {
      &:before {
        border-color: #00a2df;
        background: #10b2ef;
      }
      &:after {
        left: 21px;
      }
    }
  }
  &.semi-checked {
    + label {
      &:before {
        border-color: #ddd;
        background: #ddd;
      }

      &:after {
        content: '';
        box-sizing: content-box;
        box-sizing: initial;
        position: absolute;
        left: 6px;
        top: 10px;
        background: #999;
        width: 10px;
        height: 2px;
      }
    }
  }
}
