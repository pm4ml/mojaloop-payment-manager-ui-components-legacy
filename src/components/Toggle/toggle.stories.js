/* eslint no-alert: "off" */
/* eslint no-console: "off" */
/* eslint no-console: "off" */
import React from 'react';
import Toggle from './Toggle';
import Column from '../Column';

export default {
  title: 'Toggle',
  component: Toggle,
};

export const Default = () => (
  <Column align="space-between">
    <div className="m5">
      <Toggle label="Default Toggle" checked={false} />
    </div>
    <div className="m5">
      <Toggle label="Default Toggle" checked />
    </div>
  </Column>
);

export const Disabled = () => (
  <Column align="space-between">
    <div className="m5">
      <Toggle label="Disabled Toggle" checked={false} disabled />
    </div>
    <div className="m5">
      <Toggle label="Disabled Toggle" checked disabled />
    </div>
  </Column>
);

export const OnChange = () => (
  <Column align="space-between">
    <div className="m5">
      <Toggle label="onChange event Toggle" checked={false} onChange={console.log} />
    </div>
    <div className="m5">
      <Toggle label="onChange event Toggle" checked onChange={console.log} />
    </div>
  </Column>
);

export const OnFocus = () => (
  <Column align="space-between">
    <div className="m5">
      <Toggle label="onFocus event Toggle" checked={false} onFocus={console.log} />
    </div>
    <div className="m5">
      <Toggle label="onFocus event Toggle" checked onFocus={console.log} />
    </div>
  </Column>
);

export const OnBlur = () => (
  <Column align="space-between">
    <div className="m5">
      <Toggle label="onBlur event Toggle" checked={false} onBlur={console.log} />
    </div>
    <div className="m5">
      <Toggle label="onBlur event Toggle" checked onBlur={console.log} />
    </div>
  </Column>
);
