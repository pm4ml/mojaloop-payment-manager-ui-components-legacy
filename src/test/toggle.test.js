import { mount, shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import React from 'react';

import Toggle from '../components/Toggle';

it('renders the toggle', () => {
  const wrapper = shallow(<Toggle />);
  expect(wrapper.find('div.input-toggle')).toHaveLength(1);
});

it('renders the toggle value', () => {
  const wrapper = shallow(<Toggle checked />);
  expect(wrapper.find('input[type="checkbox"]').prop('checked')).toBe(true);
});

it('renders the label', () => {
  const wrapper = shallow(<Toggle label="test-toggle" />);
  expect(wrapper.find('label')).toHaveLength(1);
  expect(wrapper.find('label').text()).toBe('test-toggle');
});

it('applies the prop className', () => {
  const wrapper = shallow(<Toggle className="test" />);
  expect(wrapper.find('.test')).toHaveLength(1);
});

it('renders the prop id', () => {
  const wrapper = shallow(<Toggle id="testToggleId" />);
  expect(wrapper.find('#testToggleId')).toHaveLength(1);
});

it('renders the disabled state', () => {
  const wrapper = shallow(<Toggle disabled />);
  expect(wrapper.find('input')).toHaveLength(1);
  expect(wrapper.find('input').prop('disabled')).toBe(true);
});

// Events

it('triggers onFocus when focused', () => {
  const mockEvent = jest.fn();
  const wrapper = mount(<Toggle onFocus={mockEvent} />);
  expect(mockEvent).not.toHaveBeenCalled();
  wrapper.find('input').simulate('focus');
  expect(mockEvent).toHaveBeenCalled();
});

it('triggers onBlur when blurred', () => {
  const mockEvent = jest.fn();
  const wrapper = mount(<Toggle onBlur={mockEvent} />);
  expect(mockEvent).not.toHaveBeenCalled();
  wrapper.find('input').simulate('blur');
  expect(mockEvent).toHaveBeenCalled();
});

it('triggers onChange when changed', () => {
  const mockEvent = jest.fn();
  const wrapper = mount(<Toggle onChange={mockEvent} />);
  expect(mockEvent).not.toHaveBeenCalled();
  wrapper.find('label').simulate('click');
  expect(mockEvent).toHaveBeenCalled();
});

it('triggers onChange when pressing Enter key', () => {
  const mockEvent = jest.fn();
  const wrapper = mount(<Toggle onChange={mockEvent} />);
  expect(mockEvent).not.toHaveBeenCalled();
  wrapper.find('input').simulate('keyDown', { key: 'Enter', keyCode: 13, which: 13 });
  expect(mockEvent).toHaveBeenCalled();
  expect(wrapper.find('input[type="checkbox"]').prop('checked')).toBe(true);
});

it('triggers onBlur when blurring', () => {
  const mockEvent = jest.fn();
  const wrapper = mount(<Toggle onBlur={mockEvent} />);
  expect(mockEvent).not.toHaveBeenCalled();
  wrapper.find('input').simulate('blur', { key: 'Tab', keyCode: 9, which: 9 });
  expect(mockEvent).toHaveBeenCalled();
});

// Snapshot

it('renders the toggle correctly when multiple props are set', () => {
  const wrapper = shallow(<Toggle checked id="test-id" />);
  expect(shallowToJson(wrapper)).toMatchSnapshot();
});
