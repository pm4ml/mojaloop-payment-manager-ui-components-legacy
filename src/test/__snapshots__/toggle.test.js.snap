// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the toggle correctly when multiple props are set 1`] = `
<div
  className="mb-input input-toggle"
>
  <input
    checked={true}
    className="input-toggle__input"
    disabled={false}
    id="test-id"
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    type="checkbox"
  />
  <label
    className="input-toggle__label--no-margin"
    htmlFor="test-id"
    onClick={[Function]}
    role="button"
  />
</div>
`;
